File: programming/longest_substring_length_of_unique_characters/longest_substring_length_of_unique_characters.dart
import 'dart:math';

void main(List<String> args) {
  print(longestSubstringLengthOfUniqueCharacters('pwwkew'));
}

int longestSubstringLengthOfUniqueCharacters(String input) {
  var charSet = Set<String>();
  var longest = 0;
  var start = 0;

  for (var i = 0; i < input.length; i++) {
    var char = input[i];
    while (charSet.contains(char)) {
      charSet.remove(input[start]);
      start++;
    }
    charSet.add(char);
    longest = max(longest, i - start + 1);
  }

  return longest;
}

File: programming/longest_substring_length_of_unique_characters/longestSubstringLengthOfUniqueCharacters.ts
function longestSubstringLengthOfUniqueCharacters({
  input,
}: {
  input: string;
}): number {
  const charSet = new Set<string>();
  let longest = 0;
  let start = 0;

  for (let i = 0; i < input.length; i++) {
    let char = input[i];
    while (charSet.has(char)) {
      charSet.delete(input[start]);
      start++;
    }
    charSet.add(char);
    longest = Math.max(longest, i - start + 1);
  }

  return longest;
}

console.log(longestSubstringLengthOfUniqueCharacters({ input: 'pwwkew' }));
